opentelemetry-api/src/opentelemetry/metrics/instrument.py

```diff
@@ -0,0 +1,208 @@
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# pylint: disable=too-many-ancestors
+
+# type: ignore
+
+
+from abc import ABC, abstractmethod
+from logging import getLogger
+from re import compile as compile_
+from threading import RLock
+from types import GeneratorType
+
+from opentelemetry.metrics.measurement import Measurement
+
+_logger = getLogger(__name__)
+
+
+class Instrument(ABC):
+
+    _name_regex = compile_(r"[a-zA-Z][-.\w]{0,62}")
+
+    @property
+    def name(self):
+        return self._name
+
+    @property
+    def unit(self):
+        return self._unit
+
+    @property
+    def description(self):
+        return self._description
+
+    @abstractmethod
+    def __init__(self, name, *args, unit="", description="", **kwargs):
+        self._lock = RLock()
+
+        if name is None or self._name_regex.fullmatch(name) is None:
+            _logger.error("Invalid instrument name %s", name)
+
+        else:
+            self._name = name
+
+        if unit is None:
+            self._unit = ""
+        elif len(unit) > 63:
+            _logger.error("unit must be 63 characters or shorter")
+
+        elif any(ord(character) > 127 for character in unit):
+            _logger.error("unit must only contain ASCII characters")
+        else:
+            self._unit = unit
+
+        if description is None:
+            description = ""
+
+        self._description = description
+
+
+class Synchronous(Instrument):
+    pass
+
+
+class Asynchronous(Instrument):
+    @abstractmethod
+    def __init__(
+        self, name, callback, *args, unit="", description="", **kwargs
+    ):
+        super().__init__(name, *args, unit=unit, description="", **kwargs)
+        with self._lock:
+
+            if not isinstance(callback, GeneratorType):
+                _logger.error("callback must be a generator")
+
+            else:
+                super().__init__(
+                    name, unit=unit, description=description, *args, **kwargs
+                )
+                self._callback = callback
+
+    @property
+    def callback(self):
+        def function():
+            measurement = next(self._callback)
+            if not isinstance(measurement, Measurement):
+                _logger.error("Callback must return a Measurement")
+                return None
+
+            return measurement
+
+        return function
+
+
+class _Adding(Instrument):
+    pass
+
+
+class _Grouping(Instrument):
+    pass
+
+
+class _Monotonic(_Adding):
+    pass
+
+
+class _NonMonotonic(_Adding):
+    pass
+
+
+class Counter(_Monotonic, Synchronous):
+    @abstractmethod
+    def add(self, amount, attributes=None):
```

Comment:
Two questions:


1. Why is this locking if there's no shared state to protect?
2. Should this behavior be in the SDK?


If we think of the API as a 0-overhead thing that folks can use to instrument but doesn't cost anything at runtime, we should avoid (as much as possible) any kind of logic until an SDK is hooked up.

Reply:
I have removed all the locks from this PR.